File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack =[]
        mapping ={")":"(","}":"{","]":"["}
        
        for bracket in s:
            if bracket not in mapping:
                stack.append(bracket)
            elif stack and stack[-1] == mapping[bracket]:
                # 1 and 1
                # match found
                stack.pop()
            else:
                # closing bracket and stack empty (no opening bracket)
                # closing bracket and stack[-1] != mapping[bracket]
                return False
        return len(stack)==0

        
    
    # if all stack is popped when pairs come, then it will be empty. We need to account for uncompleted pairs
    
    #1 starts with end "]" stack empty and ending bracket encoutered
    #2 "[" starts with open no closing . stack not empty at the end of for loop
    #3 "(]" popped opening bracket does not match closed bracket
    #4 "{[]}" our desered result

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        Map<Character, Character> hmap = new HashMap<>();
        hmap.put(')','(');
        hmap.put(']','[');
        hmap.put('}','{');

        Stack<Character> st = new Stack<>();
        for(char ch : s.toCharArray()) {
            if (hmap.containsKey(ch)) {
                if (!st.isEmpty() &&  st.peek() == hmap.get(ch)) {
                    st.pop();
                    continue;
                }
                return false;

            }
            else {
                //opening brack add
                st.add(ch);
            }
        }
        return st.empty() ? true : false;
    }
}